Update Cholesky's forward and backward derivative

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76032

This PR:
- Derives formally a new rule for Cholesky (write-up to come)
- Implements it without using in-place operations in the forward or backward.
- Does not instantiate inverses explicitly, but rather it solves two triangular systems of equations (2 triang vs 1 triang and 2 matmuls should be comparable, but the first one should be more stable).
- Fixes almost all the batching transform / composite compliance issues these gradients had.

lezcano committed:
Update Choesky's forward and backward derivative

This PR:
- Derives formally a new rule for Cholesky (write-up to come)
- Implements it without using in-place operations in the forward or backward.
- Does not instantiate inverses explicitly, but rather it solves two triangular systems of equations (2 triang vs 1 triang and 2 matmuls should be comparable, but the first one should be more stable).

[ghstack-poisoned]

lezcano committed:
Update on "Update Choesky's forward and backward derivative"


This PR:
- Derives formally a new rule for Cholesky (write-up to come)
- Implements it without using in-place operations in the forward or backward.
- Does not instantiate inverses explicitly, but rather it solves two triangular systems of equations (2 triang vs 1 triang and 2 matmuls should be comparable, but the first one should be more stable).

[ghstack-poisoned]

lezcano committed:
Update on "Update Choesky's forward and backward derivative"


This PR:
- Derives formally a new rule for Cholesky (write-up to come)
- Implements it without using in-place operations in the forward or backward.
- Does not instantiate inverses explicitly, but rather it solves two triangular systems of equations (2 triang vs 1 triang and 2 matmuls should be comparable, but the first one should be more stable).

[ghstack-poisoned]

lezcano committed:
Update on "Update Choesky's forward and backward derivative"


This PR:
- Derives formally a new rule for Cholesky (write-up to come)
- Implements it without using in-place operations in the forward or backward.
- Does not instantiate inverses explicitly, but rather it solves two triangular systems of equations (2 triang vs 1 triang and 2 matmuls should be comparable, but the first one should be more stable).
- Fixes almost all the batched / composite issues these gradients had.

[ghstack-poisoned]

IvanYashchuk commented on torch/linalg/__init__.py:
```diff
@@ -97,9 +97,9 @@
-        skips the (slow) error checking by default and instead returns the debug
-        information. This makes it a faster way to check if a matrix is
-        positive-definite.
+        skips the error checking and its synchronization and instead returns the debug
+        information. This makes the operation faster if one knows that the decomposition
+        is going to succeed (e.g. if the input is SPD by construction).
```
You don't believe checking the return info is faster for what purpose?

This sentence was written by Mike with help from Max (https://github.com/pytorch/pytorch/pull/58217#pullrequestreview-660450116).
If your sole purpose is to check whether a matrix is a positive definite one then checking if info is zero is a lot faster than trying to catch an exception. That's the whole motivation behind those `_ex` variants. Trying to Cholesky decompose a matrix and seeing if it succeeds is the best way to tell whether the matrix is numerically positive definite. See https://github.com/pytorch/pytorch/issues/34272.

nikitaved commented on torch/linalg/__init__.py:
```diff
@@ -83,7 +83,7 @@
     A = LL^{\text{H}}\mathrlap{\qquad L \in \mathbb{K}^{n \times n}}
 
-where :math:`L` is a lower triangular matrix and
+where :math:`L` is a lower triangular matrix with real diagonal (even in the complex case) and
```
`L` should be real and positive I guess.

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1222,43 +1222,54 @@ Tensor masked_scatter_backward(const Tensor & grad, const Tensor & mask, IntArra
-  return upper ? L_tangent.mH() : L_tangent;
+  auto dL = at::linalg_solve_triangular(L_, dA, /*upper=*/false, /*left=*/true);
+  dL = at::linalg_solve_triangular(L_.mH(), dL, /*upper=*/true, /*left=*/false);
+  dL = dL.tril() - 0.5 * dL.diagonal(0, -2, -1).diag_embed();
```
I guess `(0.5 * dL.diagonal).diag_embed` should be faster.

Lezcano replied:
Good point. Even more, when https://github.com/pytorch/pytorch/pull/76115/files#r854455360 is fixed, we should be able to do it in-place again.

IvanYashchuk commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1222,43 +1222,54 @@ Tensor masked_scatter_backward(const Tensor & grad, const Tensor & mask, IntArra
+  // \pi(X) = X.tril() - 0.5*diag(X)
+  // so
+  // dL = L\pi(L^{-1}dA(L^{-H}))
+  TORCH_CHECK(at::allclose(dA, dA.mH(), /*rtol=*/1e-2, /*atol=*/1e-2),
+              "cholesky_jvp: The tangent part of the matrix A should also be ",
+              (dA.is_complex() ? "Hermitian" : "symmetric."));
```
This check is not enforced in the primal calculation and it doesn't feel right (because of performance) to include it in the forward AD rule.

Lezcano replied:
It can potentially do so if they pass a triangular matrix (cc @IvanYashchuk) but perhaps we could assume it is symmetric and roll with that?
I put it for consistency with `eigh_jvp`, but I do agree it's not the best. Should I remove it from here and also put up a PR that removes it from eigh_jvp?

nikitaved replied:
I would probably kill it, but let's ask @soulitzer whether it is needed and whether forward AD can be potentially broken by the user.

Lezcano replied:
fwiw, I currently vote for removing these checks as well. My reasoning is that the forward function needs a symmetric (and positive definite) matrix, so the user needs to map the input into a SPD matrix, in particular making the forward AD tensor `dA` symmetric. If they don't do that, even if the matrix is symmetric, it would be a symmetric matrix living in R^{n x n}, and that's not OK (you cannot expect correct forward AD / gradients in that scenario).

nikitaved replied:
Right, maybe it is worth mentioning somewhere that `A + dA` is supposed to be of the same structure as `A`...

Lezcano replied:
You probably mean that `dA` is symmetric / Hermitian (`A + dA` needn't be SPD). I've just left a comment noting this.

albanD replied:
I think the original idea behind these checks was that: "If autograd computes something, it has to be correct*".
And so if the forward/backward AD formula only works for a subset of inputs, then it should make sure to raise an error for any input that is valid for f but for which we cannot compute gradients.



* According to our own rules, notably, if the input is not valid for f, then we can do whatever.

Lezcano replied:
Offline discussion with @albanD:
The rules for autograd are:
>To try and reduce the impact of this limitation, we define the gradients of the elementary operations by applying the following rules in order:
>* If the function is not defined (:math:`\sqrt(-1)` or :math:`\log(-1)` for example) then this is undefined. Most function will return :math:`nan`, but for performance reasons, some functions will return not-nan values (:math:`\log(-1)` for example).
>* If a gradient exist at the current point, return it.
>* If the function is convex, return one of it's subgradients at the current point, if more than one exists, return the subgradient with minimum norm (as it is always a descent direction, see "Convex Optimization Algorithms" by Dimitri Bertsekas, Proposition 4.3.1 for a proof).
>* Define the gradient at the current point by continuity (note that :math:`inf` is possible here). If multiple values are possible, pick one arbitrarily.

This case falls under the first one (the function is not defined if the input matrix `dA` is not symmetric, as you're calling the `jvp` function wrong) so we can return whatever.

nikitaved commented on torch/csrc/autograd/FunctionsManual.cpp:
```diff
@@ -1222,43 +1222,54 @@ Tensor masked_scatter_backward(const Tensor & grad, const Tensor & mask, IntArra
+  // since tril(L^H triu(gL, 1)) = 0, as L^H triu(gL, 1) is upper triangular
+  auto gA = L_.mH().matmul(gL_).tril();
+  // Equivalent to 0.5 * (gA + gA^H - diag(gA))
+  gA = 0.5 * (gA + gA.tril(-1).mH());
```
nit: maybe we can do a similar trick as the one from above with `diag_embed` here as well?

lezcano committed:
Update on "Update Cholesky's forward and backward derivative"


This PR:
- Derives formally a new rule for Cholesky (write-up to come)
- Implements it without using in-place operations in the forward or backward.
- Does not instantiate inverses explicitly, but rather it solves two triangular systems of equations (2 triang vs 1 triang and 2 matmuls should be comparable, but the first one should be more stable).
- Fixes almost all the batching transform / composite compliance issues these gradients had.

[ghstack-poisoned]

nikitaved reviewed (approved):
Looks great! Thank you!

lezcano committed:
Update on "Update Cholesky's forward and backward derivative"


This PR:
- Derives formally a new rule for Cholesky (write-up to come)
- Implements it without using in-place operations in the forward or backward.
- Does not instantiate inverses explicitly, but rather it solves two triangular systems of equations (2 triang vs 1 triang and 2 matmuls should be comparable, but the first one should be more stable).
- Fixes almost all the batching transform / composite compliance issues these gradients had.

[ghstack-poisoned]

lezcano committed:
Update on "Update Cholesky's forward and backward derivative"


This PR:
- Derives formally a new rule for Cholesky (write-up to come)
- Implements it without using in-place operations in the forward or backward.
- Does not instantiate inverses explicitly, but rather it solves two triangular systems of equations (2 triang vs 1 triang and 2 matmuls should be comparable, but the first one should be more stable).
- Fixes almost all the batching transform / composite compliance issues these gradients had.

[ghstack-poisoned]

albanD commented:
The CI failure is weird... But does look unrelated.

Lezcano commented:
@pytorchbot merge this please
